### PR TITLE
Implement CA1030: Use events where appropriate

### PR DESCRIPTION
**Pull request by mavasani**
This rule detects methods that have names that ordinarily would be used for events. Events follow the Observer or Publish-Subscribe design pattern; they are used when a state change in one object must be communicated to other objects. If a method gets called in response to a clearly defined state change, the method should be invoked by an event handler. Objects that call the method should raise events instead of calling the method directly.

https://msdn.microsoft.com/en-us/library/ms182177.aspx

TODO: Implement fixer

Fixes https://github.com/dotnet/roslyn-analyzers/issues/380

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/roslyn-analysis for review.

**Comment by heejaechang** [comment]
you probably want to update props file to remove this rule from running for old FxCop?

**Comment by heejaechang** [comment]
:+1:

**Comment by heejaechang** [comment]
by the way, your analyzer also seems not checking cancellation?

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs [hunk 1]
```diff
@@ -9,12 +12,525 @@ public class UseEventsWhereAppropriateTests : DiagnosticAnalyzerTestBase
...
+        {
+            VerifyCSharp(@"
+using System;
+
+public class InterfaceImplementation : I
+{
+    // Explicit interface implementation - Rule does not fire.
+    void I.FireOnSomething_InterfaceMethod1()
+    {
+        throw new NotImplementedException();
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: The test code can be a few lines shorter if you omit all the throws.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs [hunk 2]
```diff
@@ -9,12 +12,525 @@ public class UseEventsWhereAppropriateTests : DiagnosticAnalyzerTestBase
...
+#pragma warning restore CA1030
+");
+
+            VerifyBasic(@"
+Imports System
+
+Public Class InterfaceImplementation
+	Inherits I
+	' Explicit interface implementation - Rule does not fire.
+	Private Sub I_FireOnSomething_InterfaceMethod1() Implements I.FireOnSomething_InterfaceMethod1
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: For this to be an actual test of the analyzer, wouldn't you have to remove the prefix `I_`? This name doesn't start with "fire".

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs [hunk 3]
```diff
@@ -9,12 +12,525 @@ public class UseEventsWhereAppropriateTests : DiagnosticAnalyzerTestBase
...
+Public Interface I
+	' Interface methods - Rule fires.
+	Sub FireOnSomething_InterfaceMethod1() Implements I.FireOnSomething_InterfaceMethod2
+	Sub FireOnSomething_InterfaceMethod2()
+End Interface
+#Enable Warning CA1030
+");
+        }
+
+        [WorkItem(380, "https://github.com/dotnet/roslyn-analyzers/issues/380")]
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: Why is this flagged with `WorkItem`? I thought we only did that for skipped tests.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs [hunk 4]
```diff
@@ -9,12 +12,525 @@ public class UseEventsWhereAppropriateTests : DiagnosticAnalyzerTestBase
...
+        [Fact]
+        public void NoDiagnostic_UnflaggedMethodKinds()
+        {
+            VerifyCSharp(@"
+using System;
+
+public class FireOnSomethingDerivedClass : BaseClass
+{
+    // Constuctor - Rule does not fire.
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: "Constuctor" => "Constructor" (also on Line 205)

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs [hunk 5]
```diff
@@ -9,12 +12,525 @@ public class UseEventsWhereAppropriateTests : DiagnosticAnalyzerTestBase
...
+
+Public Class FireOnSomethingDerivedClass
+	Inherits BaseClass
+	' Constuctor - Rule does not fire.
+	Public Sub New()
+	End Sub
+
+	' Finalizer - Rule does not fire.
+	Protected Overrides Sub Finalize()
+		Try
```
Comment (lgolding) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: Similar to above comment -- you don't need the method body to write the test.

**Comment by lgolding** [comment]
:+1:

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs (hunk 1 above)
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: Thanks, I'll take care of this for next PR onwards.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs (hunk 2 above)
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: Yes, thanks updated.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs (hunk 3 above)
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: For Roslyn, we did both - tests for new features are generally tagged with the work item tracking the feature work and then if a test gets broken with bug(s), we replace the work item with the bug number.

File: src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs (hunk 4 above)
Comment (mavasani) [inline src/Microsoft.ApiDesignGuidelines.Analyzers/UnitTests/UseEventsWhereAppropriateTests.cs]: Thanks, fixed.

**Comment by mavasani** [comment]
@heejaechang - I have updated the props file. I don't think checking cancellation for this analyzer is helpful as it is not doing any expensive operation and there is no loop. Driver checks for cancellation before making a callback to analyzer, so we should be fine.

However, I think we need to go through all our analyzers and audit them for cancellation checks, props file, etc. I'll open a work item for the same.

**Comment by mavasani** [comment]
Jenkins machine seem to be all down - Tests are passing locally, so merging it in.

**Pull request merged**